vega_sim/scenario/common/agents.py

```diff
@@ -1473,6 +1475,301 @@ def _calculate_price_volume_levels(
+            )
+
+    def _update_state(self, current_step: int):
+        super()._update_state(current_step)
+
+        if self.state_update_freq and current_step % self.state_update_freq == 0:
+
+            int_market_info = (
+                self.vega.market_info(market_id=self.market_id)
+                if self.market_id is not None
+                else None
+            )
+            self.int_mkr_fee = (
+                float(int_market_info.fees.factors.maker_fee)
+                if int_market_info is not None
+                else 0
+            )
+            self.int_liq_fee = (
+                float(int_market_info.fees.factors.liquidity_fee)
+                if int_market_info is not None
+                else 0
+            )
+            self.int_inf_fee = (
+                float(int_market_info.fees.factors.infrastructure_fee)
+                if int_market_info is not None
+                else 0
+            )
+
+            ext_market_info = (
+                self.vega.market_info(market_id=self.external_market_id)
+                if self.external_market_id is not None
+                else None
+            )
+            self.ext_mkr_fee = (
+                float(ext_market_info.fees.factors.maker_fee)
+                if ext_market_info is not None
+                else 0
+            )
+            self.ext_liq_fee = (
+                float(ext_market_info.fees.factors.liquidity_fee)
+                if ext_market_info is not None
+                else 0
+            )
+            self.ext_inf_fee = (
+                float(ext_market_info.fees.factors.infrastructure_fee)
+                if ext_market_info is not None
+                else 0
+            )
+
+    def _optimal_strategy(self, current_position, current_step):
+
+        ext_best_bid, ext_best_ask = self.vega.best_prices(
+            market_id=self.external_market_id
+        )
+
+        int_fee = self.int_mkr_fee + self.int_liq_fee
+        ext_fee = self.ext_mkr_fee + self.ext_liq_fee + self.ext_inf_fee
+
+        required_bid_price = (
+            ext_best_bid * (1 - ext_fee) / (1 - int_fee + self.profit_margin)
+        )
+        required_ask_price = (
+            ext_best_ask * (1 + ext_fee) / (1 + int_fee - self.profit_margin)
+        )
+
+        current_bid_depth = self.curr_price - required_bid_price
+        current_ask_depth = required_ask_price - self.curr_price
+
+        return current_bid_depth, current_ask_depth
+
+    def _balance_positions(self):
+
+        # Determine the delta between the position on the internal and external market
+
+        internal_position = self.vega.positions_by_market(
+            wallet_name=self.wallet_name,
+            market_id=self.market_id,
+            key_name=self.key_name,
+        )
+        current_int_position = (
+            float(internal_position[0].open_volume) if internal_position != [] else 0
+        )
+        external_position = self.vega.positions_by_market(
```

Comment:
I would be nice to generify the `positions_by_market` function at some point to be able to pass `None` as the market_id and get everything back. Some of our later data access functions handle this. Doesn't need doing now though, two calls isn't too bad :)

Reply:
Sounds good, raised an [issue](https://github.com/vegaprotocol/vega-market-sim/issues/285) so I don't forget 😄